// For documentation on protocol buffers, see:
//   https://developers.google.com/protocol-buffers/docs/proto

message WaveletCompressedImage {
  required int32 width = 1;
  required int32 height = 2;
  required int32 wavelet_transform_step_count = 3;

  enum WaveletAlgorithm {
    HAAR = 1;
    CDF79 = 2;
    CDF53 = 3;
  }
  required WaveletAlgorithm wavelet_algorithm = 4;

  // standard or nonstandard transform
  // standard: do all transform steps, then transpose
  // nonstandard: do one transform step, transpose, repeat
  optional bool standard_transpose = 5 [default = false];

  // This proportion of the original data that was replaced with zero.
  // It will be a value between 0 and 1.
  // (this value is not needed when reconstructing the data)
  optional float threshold_fraction = 6;

  // All values with an absolute value less than this were replaced with zero.
  optional float threshold_value = 7;

  required int32 quantize_bits = 8;

  enum QuantizationAlgorithm {
    UNIFORM = 1;
    LOG = 2;
    COUNT = 3;
    LLOYD = 4;
  }
  optional QuantizationAlgorithm quantization_algorithm = 9;

  // the maximum value used when computing quantization
  optional float quant_max_value = 10;

  // The boundaries of the quantization bins
  // For N quantizeBits, there are 2^N bins, and (2^N)+1 boundaries.
  repeated float quant_bin_boundaries = 11 [packed=true];

  // The values that should be used to represent each quantized value.
  // AKA the codebook.
  // For N quantizeBits, there are 2^N quantize values.
  repeated float quant_bin_values = 12 [packed=true];

  repeated int32 huffman_encode_table = 13 [packed=true];
}


message CubeletBuffer {
  // dimensions of this cubelet
  required uint32 width = 1;
  required uint32 height = 2;
  required uint32 depth = 3;

  // (x,y,z) position of this cubelet in the overall data set
  optional uint32 x_offset = 4 [default = 0];
  optional uint32 y_offset = 5 [default = 0];
  optional uint32 z_offset = 6 [default = 0];

  enum DataType {
    UINT8 = 1;   // 1-byte unsigned int, or raw bytes
    FLOAT32 = 2; // 4-byte float
  }
  required DataType data_type = 7;

  // number of bytes used by the data
  required uint32 byte_count = 8;

  enum CompressionAlgorithm {

    NONE = 0;   // no compression

    ZLIB = 1;   // like gzip, see https://en.wikipedia.org/wiki/Zlib

    WAVELET = 2;  // This is our custom wavelet transform->threshold->
                  // quantize->huffman compression algorithm.
                  // There are a number of parameters needed to decompress
                  // the data, so if this is the algorithm used, then
                  // those parameters will need to be saved somehow, perhaps
                  // in a WaveletCompressedImage message.
                  
  }
  optional CompressionAlgorithm compression_algorithm = 9 [default = NONE];

  optional uint32 checksum = 10;

  // If the data corresponding to this Cubelet message is not immediately
  // adjacent to the message metadata, then data_offset and either
  // data_file_index or data_file_name should be set.

  // Only one of data_file_name or data_file_index should be set.
  // If data_file_name is set, it should be a relative reference
  // (mycubelets.data or data/cubelets.0) not an absolute reference
  // (c:\data\cubelets.data or /home/bob/data/cubes.0) so the index
  // file and data file can be moved to another location.

  // If there are many different data files in this set of cubelets, then
  // store the name of each data file with each cubelet in data_file_name.
  // If there not many different data files, store the file names in an
  // array (for example, CubeletIndex::data_files) and reference the
  // data file using data_file_index.
  optional string data_file_name = 11;
  optional uint32 data_file_index = 12;

  // Offset from the beginning in the data file where the data for this
  // cubelet can be found.
  optional uint64 data_file_offset = 13;
}


message CubeletIndexBuffer {
  // overall size of the object
  optional uint32 width = 1;
  optional uint32 height = 2;
  optional uint32 depth = 3;

  // array of data file names
  repeated string data_files = 4;

  // array of cubelet metadata objects
  repeated CubeletBuffer cubelets = 5;
}
